doc/develguide.rst

```diff
@@ -0,0 +1,656 @@
+Developer Guide
+===============
+
+Overview
+--------
+
+This document describes the guts of jpype.  It is intended 
+lay out the architecture of the jpype code to aid intrepid lurkers 
+to develop and debug the jpype code once I am run over by a bus.
+
+
+History
+~~~~~~~
+
+When I started work on this project it had already existed for over 10 years.
+The original developer had intended a much larger design with modules to
+support multiple languages such as Ruby.  As such it was constructed with 
+three layers of abstraction.  It has a wrapper layer over Java in C++, a
+wrapper layer for the Python api in C++, and an abstraction layer intended
+to bridge Python and other interpreted languages.  This multilayer abstraction
+ment that every debugging call had to drop through all of those layers.
+Memory management was split into multiple pieces with Java controlling a
+portion of it, C++ holding a bunch of resources, Python holding additional
+resources, and HostRef controlling the lifetime of objects shared between the
+layers.  It also had its own reference counting system for handing Java
+references on a local scale.
+
+This level of complexity was just about enough to scare off all but the most
+hardened programmer.  Thus I set out to eliminate as much of this as I could.
+Java already has its own local referencing system to deal in the form of 
+LocalFrames.  It was simiply a matter of setting up a C++ object to 
+hold the scope of the frames to eliminate that layer.  The Java abstraction
+was laid out in a fashion somewhat orthagonally to the Java inheritance 
+diagram.  Thus that was reworked to something more in line which could be 
+safely completed without disturbing other layers.  The multilanguage 
+abstraction layer was already perced in multiple ways for speed.  However, 
+as this interwove throughout all the kit it was a terrible lift to remove
+and thus required gutting the Python layer as well to support the operations
+that were being performed by the HostRef.  
+
+The remaining codebase is fairly slim and reasonably streamlined.  This 
+rework cut out about 30% of the existing code and sped up the internal
+operations.  The Java C++ interface matches the Java class hierachy.
```

Comment:
i really like your history section! It gives me a great overview of what you have achieved.

Reply:
Thanks.